service/storage_service.cc

```diff
@@ -516,7 +520,21 @@ void storage_service::prepare_to_join(std::vector<inet_address> loaded_endpoints
     bool restarting_normal_node = db::system_keyspace::bootstrap_complete() && !db().local().is_replacing() && !my_tokens.empty();
     if (restarting_normal_node) {
         slogger.info("Restarting a node in NORMAL status");
+        // This node must know about its chosen tokens before other nodes do
+        // since they may start sending writes to this node after it gossips status = NORMAL.
+        // Therefore we update _token_metadata now, before gossip starts.
         _token_metadata.update_normal_tokens(my_tokens, get_broadcast_address());
+
+        _cdc_streams_ts = db::system_keyspace::get_saved_cdc_streams_timestamp().get0();
+        if (!_cdc_streams_ts && db().local().get_config().check_experimental(db::experimental_features_t::CDC)) {
+            // We could not have completed joining if we didn't generate and persist a CDC streams timestamp,
+            // unless we are restarting after upgrading from non-CDC supported version.
+            // In that case we won't begin a CDC generation: it should be done by one of the nodes
+            // after it learns that it everyone supports the CDC feature.
+            cdc_log.warn(
+                    "Restarting node in NORMAL status with CDC enabled, but no streams timestamp was proposed"
+                    " by this node according to its local tables. Are we upgrading from a non-CDC supported version?");
+        }
     }
```

Comment:
Let's make this a call into the cdc module to avoid spreading the cdc code everywhere.

Reply:
The pieces of code that I added inside `join_token_ring` and `prepare_to_join` are, I believe, too entangled with the node startup code to move them out of context.

The only work this particular does is
```
_cdc_streams_ts = db::system_keyspace::get_saved_cdc_streams_timestamp().get0();
````
I don't see a suitable abstraction for this. The rest is just a bunch of comments explaining why this line has to be here and not in a different place